### ISSUE TITLE
Surprising random seed behavior when using @tf.function

### ISSUE THREAD

**Issue by ageron** [tensorflow/tensorflow#33297]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOSX 10.14.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d382ca 2.0.0
- Python version: 3.7.4
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**

Random seeds work in surprising ways in TF 2.0 when using `@tf.function`. In particular, the value of the global random seed is only taken into account when a function is traced, not when it is called. This is surprising and different from TF 1.x behavior.

**Describe the expected behavior**
I expect the value of the global random seed to be taken into account every time a pseudo-random number is generated.

**Code to reproduce the issue**

```python
@tf.function
def rnd():
    return tf.random.uniform(shape=[])

tf.random.set_seed(42)
print(rnd()) # The rnd() function's seed is generated randomly now, based on
print(rnd()) # the current random seed (which is 42).
print()

tf.random.set_seed(43) # resets the random sequence but ignores this seed!
print(rnd())
print(rnd())
print()

tf.random.set_seed(42) # resets the random sequence but ignores this seed!
print(rnd())
print(rnd())
print()
```

The output value is:

```
tf.Tensor(0.63789964, shape=(), dtype=float32)
tf.Tensor(0.8774011, shape=(), dtype=float32)

tf.Tensor(0.63789964, shape=(), dtype=float32)
tf.Tensor(0.8774011, shape=(), dtype=float32)

tf.Tensor(0.63789964, shape=(), dtype=float32)
tf.Tensor(0.8774011, shape=(), dtype=float32)
```

Notice that we get the same sequence of random numbers every time, ignoring the value of the global random seed. The only value that matters is the first one (when the function gets traced).

More code and examples of surprising behavior in [this colab](https://colab.research.google.com/drive/1C3LZkt5hfO6T2Uo2xaYVG8hiNQL8c3xu).

**Other info / logs**

Specifically, I would expect the output to look the same as when the function is not decorated with `@tf.function`:

```
tf.Tensor(0.6645621, shape=(), dtype=float32)
tf.Tensor(0.68789124, shape=(), dtype=float32)

tf.Tensor(0.2733041, shape=(), dtype=float32)
tf.Tensor(0.5168259, shape=(), dtype=float32)

tf.Tensor(0.6645621, shape=(), dtype=float32)
tf.Tensor(0.68789124, shape=(), dtype=float32)
```

Note that the second sequence is different, as expected (in fact, the pseudo-random numbers should be identical whether the function is decorated or not, but that's a nice-to-have).

**Comment by rmothukuru** [comment]
@ageron,
Can you please refer the issue #33034 and let us know if it helps. Thanks!

**Comment by ageron** [comment]
Hi @rmothukuru ,
I looked at issue #33034 before filing this issue, I think it's a different problem. I am not using per-operation seeds in this issue.

**Comment by rmothukuru** [comment]
Could reproduce the issue with TF Version 2.0. Here is the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/faae8f0424cd0bcfcee29854140cab8c/random-seeds-in-tf-2-0.ipynb).

**Comment by mdanatg** [comment]
@alextp @wangpengmit FYI

Unfortunately, this is a known issue of `tf.random.seed`. We'll fix the documentation to help avoid the issue, and I'll post an example for how the original example can be updated so that it consistently works in `tf.function` as well..

**Comment by mdanatg** [comment]
Update: here is a method that currently works in TF2, and which will give you the expected results. Unfortunately, `tf.set_seed` and `tf.random.uniform` (and similar ops) are currently broken, although they will probably be either fixed or removed in the future:

```
@tf.function
def rnd(rng):
  return rng.uniform(shape=[])

rng = tf.random.experimental.Generator.from_seed(seed=42)
print(rnd(rng))
print(rnd(rng))
print()

rng = tf.random.experimental.Generator.from_seed(seed=43)
print(rnd(rng))
print(rnd(rng))
print()

rng = tf.random.experimental.Generator.from_seed(seed=42)
print(rnd(rng))
print(rnd(rng))
print()
```

Note: currently, calling the function with a new generator will causing it to be retraced (that is, will be slow the first time you call it); this will be fixed soon.

@yashk2810 it would be nice to make sure this snippet or something similar is visible in one of our tutorials or perhaps the docs for tf.random?

**Comment by yashk2810** [comment]
Its fixed now: 
* https://www.tensorflow.org/api_docs/python/tf/random/set_seed
* https://www.tensorflow.org/api_docs/python/tf/random/uniform

**Comment by yashk2810** [comment]
We also link off to the notebooks where the particular symbol is being used.

See this for example: https://www.tensorflow.org/api_docs/python/tf/random/uniform#used_in_the_guide

**Comment by ageron** [comment]
Thanks @mdanatg and @yashk2810 , the documentation is clearer now, and it's nice to have a workaround. I really like the `tf.random.experimental.Generator` solution, and in fact I think I'll stop using `tf.random.set_seed()` altogether, as well as op-level seeds. Their behavior is way too complicated and error-prone, IMHO.

Ideally, a function's behavior should not change when you decorate it with `@tf.function`, and things should be as simple and intuitive as possible. In NumPy, there is a single global random number generator (RNG), and you can create new ones if you need them. All random ops use it by default. If users need other RNGs, they just create them. Works fine.

Here's a solution that works fine and which I find very simple and intuitive:

```python
import tensorflow as tf

tf.random_v2 = tf.random.experimental.Generator.from_non_deterministic_state()
tf.random_v2.set_seed = tf.random_v2.reset_from_seed

@tf.function
def rng():
    return tf.random_v2.uniform(shape=[])

print(rng()) # prints ?1, different for every run of this program
print(rng()) # prints ?2

tf.random_v2.set_seed(42)
print(rng()) # prints A1
print(rng()) # prints A2

tf.random_v2.set_seed(43)
print(rng()) # prints B1
print(rng()) # prints B2

tf.random_v2.set_seed(42)
print(rng()) # prints A1
print(rng()) # prints A2
```

Everything works exactly like you would expect, and it gives the exact same result whether the `rng()` function is decorated with `@tf.function` or not.

The only problem is that `reset_from_seed()` does not support `None`. That's really something that should be fixed IMHO, it should generate a random state like in `from_non_deterministic_state(), using `non_deterministic_ints()`.

Next, if I want to use a specific RNG, I can do so easily:

```python
@def my_fn(rng):
    return rng.uniform(shape=[])

rng1 = tf.random.experimental.Generator.from_seed(42)
rng2 = tf.random.experimental.Generator.from_seed(43)

print(my_fn(rng1)) # prints A1
print(my_fn(rng1)) # prints A2

print(my_fn(rng2)) # prints B1
print(my_fn(rng2)) # prints B2

print(my_fn(rng1)) # prints A3
print(my_fn(rng1)) # prints A4

rng1.set_seed(42)
print(my_fn(rng1)) # prints A1
print(my_fn(rng1)) # prints A2
``` 

Again, this is simple and behaves exactly like you would expect, and just the same way with or without decorating the function with `@tf.function`.

It's a pity the `reset()` method explodes if I don't pass it a `state` argument. I would expect it to revert to its initial state instead. It should just save its initial state upon creation, and revert to that state when `reset()` is called without any argument.

Wdyt?

**Comment by mdanatg** [comment]
Thank you for the thorough suggestion!

Yes, it's absolutely crucial that code works the same way with or without `tf.function` - anything that doesn't is a bug.

I'll let Peng weigh in on a solution, he's the RNG expert. Personally I like your proposal because it's clean and minimizes the symbols we'd need to export: `Generator` is the sole RNG, the "easy" `tf.random.uniform` functions just offer access to the default generator and we can omit `tf.stateless.stateless_uniform`.

**Comment by ageron** [comment]
One thing I'm not sure about is how to add this to TF 2.0 while keeping backward compatibility. I see a few options:

* Keep the existing `tf.random` and add `tf.random_v2` in parallel, as I did. But it feels wrong to add yet a *_v2 just a few weeks after 2.0 is released, which had finally removed all the *_v2s that had accumulated in TF1.
* Add `tf.random.enable_simple_behavior()` to make the new behavior opt-in. Not very elegant: shouldn't the simple behavior be the default? Perhaps make it the default behavior in TF 2.1?
* Or just update `tf.random` to use the new behavior and consider that we're just fixing a bug. Since the output differs when you decorate the function with `@tf.function`, it is a bug. Since TF 2 was just released, there's probably not much production code out there anyway, and the impact is limited: it will just change the random sequences. But in that case the recently added documentation should be updated quickly.

**Comment by wangpengmit** [comment]
Hi @ageron, your code is roughly what we have in mind for the future `tf.random.uniform` and alike, if we choose to keep them. `Generator.reset_from_non_deterministic_state` seems like a useful feature which I'll consider. I'm hesitant to make `Generator` save the initial state to support an argument-less `reset`, since for basic facilities such `Generator` we need to be careful about memory usage. After all, the user can retrieve and store the initial state herself and use it to reset the generator later.

**Comment by ageron** [comment]
Hi @wangershi ,

Thanks for your feedback. Just to be clear, are you planning to go for option 3 in my previous comment? In other words, will the new (simpler) behavior replace the old one or will they live side by side (if so, how)?

Also, I'd like to suggest providing a `set_seed()` method in the `Generator` class that calls `reset_from_seed()` if a seed is provided or `reset_from_non_deterministic_state()` if not. IMHO, it would be more convenient, more intuitive and less verbose.

Regarding the initial state, I see your point, it makes sense. Perhaps the initial state could be saved by default when creating the `Generator` (since most people won't create millions of `Generator`s, RAM usage should be fine) and have an option in the constructor such as `save_initial_state=True` so that people who need the extra RAM can set this to `False` if they want? Or perhaps set it to `False` by default and make the `reset()` method print our a clear error message if the user doesn't provide a state: "You need to set save_initial_state=True when constructing the Generator if you want to be able to call reset without any argument". Not sure about this.

**Comment by wangershi** [comment]
@ageron Your @ not me :)

**Comment by wangpengmit** [comment]
We are pretty sure the old behavior is wrong and will go away. We haven't decided whether to remove symbols such as `tf.random.uniform` and only provide `tf.random.Generator`, or to give `tf.random.uniform` the new behavior. Thank you for the other suggestions! We'll consider them.

**Comment by ageron** [comment]
Thanks for your feedback. Please don't remove `tf.random.uniform()` and other random functions: there's really no need to break everyone's existing code, IMHO. Most people won't even notice the change. I think this modification is mostly an implementation detail and a bug fix, it doesn't justify changing the API much. Just adding a warning in the release notes explaining what's changed and the motivation would be sufficient.

Moreover, I would recommend making the default `Generator` available directly through the API. Something like `tf.random.default_generator()` or `tf.random.get_default_generator()`, for example in case someone needs to pass a `Generator` to a function, and they want to pass the default one.

Thanks for your work, it's a really nice improvement!

**Comment by wangpengmit** [comment]
We already exported `tf.random.experimental.get_global_generator` (https://github.com/tensorflow/tensorflow/blob/ba96c40cb452c68ba08eb30f4e34795f4774cb92/tensorflow/python/ops/stateful_random_ops.py#L682). Is it the same as the `tf.random.get_default_generator` you have in mind?

**Comment by ageron** [comment]
Ah yes, that's exactly what I had in mind, thanks @wangpengmit . 👍

**Comment by ageron** [comment]
Hey @wangpengmit ,
When do you expect `tf.random.experimental.get|set_global_generator()` functions to be moved to `tf.random`, and when will `tf.random.uniform` (and other random functions) move to using the global generator by default?

**Comment by wangpengmit** [comment]
We don't have a timeline yet.

**Comment by ageron** [comment]
I understand. In the meantime, don't you want to leave this issue open? The `Generator` class is a great workaround, but as long as the default random functions don't use it, they're still kind of broken, IMHO.

**Comment by wangpengmit** [comment]
Sure, I've reopened the issue.

**Comment by wangpengmit** [comment]
Hi @ageron, there is a discussion about RNG migration plan: https://groups.google.com/a/tensorflow.org/d/topic/developers/WLBbnQMB2Rw/discussion . We welcome your comments!

**Comment by ageron** [comment]
Thanks @wangpengmit ,
It looks good to me. I added a note about making it possible to preserve the old behavior if possible, but I understand if that's hard to do.
